crowdnotifier-sdk/src/main/java/org/crowdnotifier/android/sdk/utils/CryptoUtils.java

```diff
@@ -68,68 +71,156 @@ public EncryptedVenueVisit getEncryptedVenueVisit(long arrivalTime, long departu
+	public byte[] decryptInternal(EncryptedData encryptedData, G1 secretKeyForIdentity, byte[] identity) {
+		G2 c1 = new G2();
+		c1.deserialize(encryptedData.getC1());
 
+		GT gt_temp = new GT();
+		Mcl.pairing(gt_temp, secretKeyForIdentity, c1);
 
-			if (!Arrays.equals(venuePrivateKey, eventInfo.getSecretKey())) {
-				continue;
-			}
+		byte[] hash = crypto_hash_sha256(gt_temp.serialize());
+		byte[] x_p = xor(encryptedData.getC2(), hash);
 
-			Payload payload = new Gson().fromJson(new String(aux), Payload.class);
+		byte[] msg_p = new byte[encryptedData.getC3().length - Box.MACBYTES];
+		int result = sodium.crypto_secretbox_open_easy(msg_p, encryptedData.getC3(), encryptedData.getC3().length,
+				encryptedData.getNonce(), crypto_hash_sha256(x_p));
+		if (result != 0) return null;
 
-			byte[] encryptedMessage = eventInfo.getEncryptedMessage();
-			byte[] decryptedMessage = new byte[encryptedMessage.length - Box.MACBYTES];
-			result = sodium.crypto_secretbox_open_easy(decryptedMessage, encryptedMessage, encryptedMessage.length,
-					eventInfo.getNonce(), payload.getNotificationKey());
-			if (result != 0) { throw new RuntimeException("crypto_secretbox_open_easy returned a value != 0"); }
+		//Additional verification
+		Fr r_p = new Fr();
+		r_p.setHashOf(concatenate(x_p, concatenate(identity, msg_p)));
 
-			String decryptedMessageString = new String(decryptedMessage);
+		G2 c1_p = new G2();
+		Mcl.mul(c1_p, baseG2(), r_p);
 
-			ExposureEvent exposureEvent = new ExposureEvent(venueVisit.getId(), payload.getArrivalTime(),
-					payload.getDepartureTime(), decryptedMessageString);
+		if (!c1.equals(c1_p)) {
+			return null;
+		}
 
-			exposureEvents.add(exposureEvent);
+		if (!secretKeyForIdentity.isValidOrder() || secretKeyForIdentity.isZero()) {
+			return null;
 		}
-		return exposureEvents;
+		return msg_p;
+	}
+
+	public EncryptedData encryptInternal(byte[] message, byte[] identity, G2 masterPublicKey) {
+
+		byte[] nonceX = randombytes_buf();
+
+		Fr r = new Fr();
+		r.setHashOf(concatenate(nonceX, concatenate(identity, message)));
+
+		G2 c1 = new G2();
+		Mcl.mul(c1, baseG2(), r);
+
+		G1 g1_temp = new G1();
+		Mcl.hashAndMapToG1(g1_temp, identity);
+
+		GT gt1_temp = new GT();
+		Mcl.pairing(gt1_temp, g1_temp, masterPublicKey);
+
+		GT gt_temp = new GT();
+		Mcl.pow(gt_temp, gt1_temp, r);
+
+		byte[] c2_pair = crypto_hash_sha256(gt_temp.serialize());
+		byte[] c2 = xor(nonceX, c2_pair);
+
+		byte[] nonce = randombytes_buf();
+
+		byte[] c3 = crypto_secretbox_easy(crypto_hash_sha256(nonceX), message, nonce);
+
+		return new EncryptedData(c1.serialize(), c2, c3, nonce);
+	}
+
+	public byte[] generateIdentity(Qr.QRCodeContent qrCodeContent, byte[] nonce1, byte[] nonce2, int hour) {
+		byte[] hash1 = crypto_hash_sha256(concatenate(qrCodeContent.toByteArray(), nonce1));
+		return crypto_hash_sha256(concatenate(hash1, concatenate(String.valueOf(hour).getBytes(), nonce2)));
```

Comment:
This is not consistent with https://github.com/CrowdNotifier/libcrowdnotifier/blob/8f25cb630cab325fe60f9b558fa109bc00967ad2/lib/src/v2/helpers.ts#L61 that does `hash1 || nonce2 || counter`